lib/image-stream/gzip.js

```diff
@@ -38,37 +35,34 @@ const ISIZE_LENGTH = 4;
  * by reading its `ISIZE`. The specification clarifies that this
  * value is just an estimation.
  *
- * @param {String} file - path to gzip file
+ * @param {Object} options - options
+ * @param {Number} options.size - file size
+ * @param {Function} options.read - read function (count, offset)
  * @fulfil {Number} - uncompressed size
  * @returns {Promise}
  *
  * @example
- * gzip.getUncompressedSize('path/to/file.gz').then((uncompressedSize) => {
+ * const fd = fs.openSync('path/to/image', 'r');
+ *
+ * gzip.getUncompressedSize({
+ *   size: fs.statSync('path/to/image.gz').size,
+ *   read: (count, offset) => {
+ *     const buffer = Buffer.alloc(count);
+ *     fs.readSync(fd, buffer, 0, count, offset);
+ *     return buffer;
+ *   }
+ * }).then((uncompressedSize) => {
  *   console.log(`The uncompressed size is: ${uncompressedSize}`);
+ *   fs.closeSync(fd);
  * });
  */
-exports.getUncompressedSize = (file) => {
-  return Bluebird.using(fs.openAsync(file, 'r').disposer((fileDescriptor) => {
-    return fs.closeAsync(fileDescriptor);
-  }), (fileDescriptor) => {
-    return fs.fstatAsync(fileDescriptor).then((stats) => {
-      const ISIZE_BUFFER_FILL_VALUE = 0;
-      const ISIZE_BUFFER_START = 0;
-      const isizeBuffer = Buffer.alloc(ISIZE_LENGTH, ISIZE_BUFFER_FILL_VALUE);
-
-      return fs.readAsync(
-        fileDescriptor,
-        isizeBuffer,
-        ISIZE_BUFFER_START,
-        ISIZE_LENGTH,
-        stats.size - ISIZE_LENGTH
-      ).then((bytesRead) => {
-        if (bytesRead !== ISIZE_LENGTH) {
-          throw new Error(`Bytes read mismatch: ${bytesRead} != ${ISIZE_LENGTH}`);
-        }
+exports.getUncompressedSize = (options) => {
+  const ISIZE_BUFFER_START = 0;
 
-        return isizeBuffer.readUInt32LE(ISIZE_BUFFER_START);
-      });
-    });
+  // Determine the uncompressed size of the gzip file by reading
+  // its `ISIZE`. The specification clarifies that this value is
+  // just an estimation.
+  return options.read(ISIZE_LENGTH, options.size - ISIZE_LENGTH).then((buffer) => {
```

Comment:
Only a minor nit, but I wonder if adding a `const ISIZE_OFFSET = options.size - ISIZE_LENGTH;` line here would help to clarify this code? (or perhaps such a small change isn't worth it)